strawberryfields/gbs/similarity.py

```diff
@@ -201,3 +209,145 @@ def event_to_sample(photon_number: int, max_count_per_mode: int, modes: int) ->
 
     return sample
+
+
+def orbit_cardinality(orbit: list, modes: int) -> int:
+    """Gives the number of samples belonging to the input orbit.
+
+    **Example usage**:
+
+    >>> orbit_cardinality([2, 1, 1], 4)
+    12
+
+    Args:
+        orbit (list[int]): orbit to count number of samples
+        modes (int): number of modes in the sample
+
+    Returns:
+        int: number of samples in the orbit
+    """
+
+    sample = list(orbit) + [0] * (modes - len(orbit))
+    counts = list(Counter(sample).values())
+    return int(factorial(modes) / np.prod(factorial(counts)))
+
+
+def event_cardinality(photon_number: int, max_count_per_mode: int, modes: int) -> int:
+    """Gives the number of samples belonging to the input event.
+
+    **Example usage**:
+
+    >>> event_cardinality(5, 2, 6)
+    126
+
+    Args:
+        photon_number (int): number of photons in the event
+        max_count_per_mode (int): maximum number of photons per mode in the event
+        modes (int): number of modes in the samples
+
+    Returns:
+        int: number of samples in the event
+    """
+    cardinality = 0
+
+    for orb in orbits(photon_number):
+        if max(orb) <= max_count_per_mode:
+            cardinality += orbit_cardinality(orb, modes)
+
+    return cardinality
+
+
+def p_orbit_mc(graph: nx.Graph, orbit: list, n_mean: float = 5, samples: int = 1000) -> float:
+    """Gives a Monte Carlo estimate of the probability of a given orbit with respect to a graph.
+
+    To make this estimate, several samples from the orbit are drawn uniformly at random using
+    :func:`orbit_to_sample`.
+
+    For each sample, this function calculates the probability of observing that sample from a GBS
+    device programmed according to the input graph and mean photon number. The sum of the
+    probabilities is then rescaled according to the cardinality of the orbit and the total number of
+     samples. The estimate is the sample mean of the rescaled probabilities.
+
+    **Example usage**:
+
+    >>> graph = nx.complete_graph(10)
+    >>> p_orbit_mc([2, 1, 1], graph)
+    0.5
+
+    Args:
+        orbit (list[int]): orbit for which to estimate the probability
+        graph (nx.Graph): input graph encoded in the GBS device
+        n_mean (float): total mean photon number of the GBS device
+        samples (int): number of samples used in the Monte Carlo estimation
+
+    Returns:
+        float: estimated orbit probability
+    """
+
+    modes = graph.order()
+    photons = np.sum(orbit)
+    A = nx.to_numpy_array(graph)
+    mean_photon_per_mode = n_mean / float(modes)
+
+    p = sf.Program(modes)
+    with p.context as q:
+        sf.ops.GraphEmbed(A, mean_photon_per_mode=mean_photon_per_mode) | q
+    eng = sf.LocalEngine(backend="gaussian")
+    result = eng.run(p)
+
+    prob = 0
+
+    for _ in range(samples):
+        sample = orbit_to_sample(orbit, modes)
+        prob += result.state.fock_prob(sample, cutoff=photons+1)

```

Comment:
What does the `cutoff` arg do?

Reply:
I'm not 100% sure but I think it determines the max size of hafnians that enter the calculations. Regardless, it will throw an error if it is set to any number smaller than `photons + 1`

Reply:
Oh 🤔 And it's not something we need to be e.g. 2 * photons?
@nquesada?

Reply:
No, the error raised said something like "cutoff must be larger than number of photons", so `photons+1` is enough